I've just tried a very quick check and I'm hitting a test failure:

```
$ nix run . -- 1
warning: creating lock file '"/home/adam/vcs/nix-heuristic-gc/flake.lock"': 
• Added input 'flake-utils':
    'github:numtide/flake-utils/11707dc2f618dd54ca8739b309ec4fc024de578b?narHash=sha256-l0KFg5HjrsfsO/JpG%2Br7fRrqm12kzFHyUHqHCVpMMbI%3D' (2024-11-13)
• Added input 'flake-utils/systems':
    'github:nix-systems/default/da67096a3b9bf56a91d16901293e51ba5b49a27e?narHash=sha256-Vy1rq5AaRuLzOxct8nz4T6wlgyUR7zLU309k9mBC768%3D' (2023-04-09)
• Added input 'nixpkgs':
    'github:NixOS/nixpkgs/3de8f8d73e35724bf9abef41f1bdbedda1e14a31?narHash=sha256-IYlYnp4O4dzEpL77BD/lj5NnJy2J8qbHkNSFiPBCbqo%3D' (2025-11-01)
error: builder for '/nix/store/vzkbxhm36qb06grky7ndgg93yybyzna8-python3.12-nix-heuristic-gc-0.6.2.drv' failed with exit code 1;
       last 25 log lines:
       > >       assert "requesting deletion of 0 store paths, total nar_size 0 bytes, 0 inodes" in caplog.text
       > E       AssertionError: assert 'requesting deletion of 0 store paths, total nar_size 0 bytes, 0 inodes' in 'INFO     nix_heuristic_gc.graph:graph.py:171 querying dead paths\nINFO     nix_heuristic_gc.graph:graph.py:192 topolo...e\nINFO     nix_heuristic_gc:__init__.py:74 (not) requesting deletion of 0 store paths, total size 0 bytes, 0 inodes\n'
       > E        +  where 'INFO     nix_heuristic_gc.graph:graph.py:171 querying dead paths\nINFO     nix_heuristic_gc.graph:graph.py:192 topolo...e\nINFO     nix_heuristic_gc:__init__.py:74 (not) requesting deletion of 0 store paths, total size 0 bytes, 0 inodes\n' = <_pytest.logging.LogCaptureFixture object at 0x7ffff0f6a6c0>.text
       >
       > tests/test_nix_heuristic_gc.py:22: AssertionError
       > ----------------------------- Captured stderr call -----------------------------
       > finding garbage collector roots...
       > determining live/dead paths...
       > ------------------------------ Captured log call -------------------------------
       > INFO     nix_heuristic_gc.graph:graph.py:171 querying dead paths
       > INFO     nix_heuristic_gc.graph:graph.py:192 topologically sorting paths
       > INFO     nix_heuristic_gc.graph:graph.py:199 building graph
       > INFO     nix_heuristic_gc.graph:graph.py:238 populating output-drv or drv-output edges
       > DEBUG    nix_heuristic_gc.graph:graph.py:262 gathering nodes for heap
       > INFO     nix_heuristic_gc.graph:graph.py:277 constructing heap
       > INFO     nix_heuristic_gc:__init__.py:70 selecting store paths for removal
       > DEBUG    nix_heuristic_gc:__init__.py:71 using limit of Quantity(value=1000000.0, unit=<QuantityUnit.INODES: 2>)
       > WARNING  nix_heuristic_gc.graph:graph.py:387 ran out of zero-reference paths to remove
       > INFO     nix_heuristic_gc:__init__.py:74 (not) requesting deletion of 0 store paths, total size 0 bytes, 0 inodes
       > =========================== short test summary info ============================
       > FAILED tests/test_nix_heuristic_gc.py::test_empty_local_store[QuantityUnit.BYTES-False] - AssertionError: assert 'requesting deletion of 0 store paths, total nar_siz...
       > FAILED tests/test_nix_heuristic_gc.py::test_empty_local_store[QuantityUnit.BYTES-True] - AssertionError: assert 'requesting deletion of 0 store paths, total nar_siz...
       > FAILED tests/test_nix_heuristic_gc.py::test_empty_local_store[QuantityUnit.INODES-False] - AssertionError: assert 'requesting deletion of 0 store paths, total nar_siz...
       > FAILED tests/test_nix_heuristic_gc.py::test_empty_local_store[QuantityUnit.INODES-True] - AssertionError: assert 'requesting deletion of 0 store paths, total nar_siz...
       > ========================= 4 failed, 1 passed in 0.11s ==========================
       For full logs, run:
         nix log /nix/store/vzkbxhm36qb06grky7ndgg93yybyzna8-python3.12-nix-heuristic-gc-0.6.2.drv
```

I assume it's straightforward to build and run the script without first running the test phase, but that's not something I'll manage to test tonight.

--- a/nix_heuristic_gc/__init__.py
+++ b/nix_heuristic_gc/__init__.py
@@ -58,11 +58,11 @@ def nix_heuristic_gc(
         penalize_exceeding_limit=_unfriendly_weight(penalize_exceeding_limit, 5e5),
     )
 
-    if garbage_graph.invalid_paths:
+    if garbage_graph.very_invalid_paths:
         logger.info(
             "Unable to handle invalid paths %s - use standard nix tools to "
             "remove these.",
-            garbage_graph.invalid_paths,
+            garbage_graph.very_invalid_paths,
         )
 
     logger.info("selecting store paths for removal")
@@ -70,11 +70,11 @@ def nix_heuristic_gc(
     to_reclaim = garbage_graph.remove_to_limit(limit.value)
 
     logger.info(
-        "%(maybe_not)srequesting deletion of %(count)s store paths, total nar_size %(size)s, %(inodes)s inodes",
+        "%(maybe_not)srequesting deletion of %(count)s store paths, total size %(size)s, %(inodes)s inodes",
         {
             "maybe_not": "(not) " if dry_run else "",
             "count": len(to_reclaim),
-            "size": format_size(sum(spn.nar_size for spn in to_reclaim), binary=True),
+            "size": format_size(sum(spn.size for spn in to_reclaim), binary=True),
             "inodes": sum(spn.inodes for spn in to_reclaim),
         },
     )

--- a/nix_heuristic_gc/__init__.py
+++ b/nix_heuristic_gc/__init__.py
@@ -28,6 +28,7 @@ def _unfriendly_weight(
 
 def nix_heuristic_gc(
     limit:Quantity,
+    penalize_invalid:int=5,
     penalize_substitutable:int=0,
     penalize_drvs:int=0,
     penalize_inodes:int=0,
@@ -56,6 +57,7 @@ def nix_heuristic_gc(
         penalize_inodes=_unfriendly_weight(penalize_inodes, 1e6),
         penalize_size=_unfriendly_weight(penalize_size, 1e-3),
         penalize_exceeding_limit=_unfriendly_weight(penalize_exceeding_limit, 5e5),
+        penalize_invalid=_unfriendly_weight(penalize_invalid, 1e6),
     )
 
     if garbage_graph.very_invalid_paths:

--- a/nix_heuristic_gc/__init__.py
+++ b/nix_heuristic_gc/__init__.py
@@ -28,7 +28,7 @@ def _unfriendly_weight(
 
 def nix_heuristic_gc(
     limit:Quantity,
-    penalize_substitutable:int=5,
+    penalize_substitutable:int=0,
     penalize_drvs:int=0,
     penalize_inodes:int=0,
     penalize_size:int=0,

--- a/nix_heuristic_gc/__main__.py
+++ b/nix_heuristic_gc/__main__.py
@@ -62,6 +62,17 @@ def main():
 
     _add_penalize_args(
         parser.add_mutually_exclusive_group(),
+        "invalid",
+        True,
+        _weighting_help_text(
+            "Prefer choosing 'invalid' paths for deletion",
+            "--penalize-invalid",
+            "Enabled by default. Invalid paths are usually the result of a failed "
+            "build and generally have no use other than debugging",
+        )
+    )
+    _add_penalize_args(
+        parser.add_mutually_exclusive_group(),
         "drvs",
         False,
         _weighting_help_text(

--- a/nix_heuristic_gc/graph.py
+++ b/nix_heuristic_gc/graph.py
@@ -28,7 +28,7 @@ class GarbageGraph:
     @dataclass(slots=True)
     class BaseStorePathNode:
         path: str
-        nar_size: int
+        nar_size: Optional[int]
         _inherited_max_atime:Optional[int] = None
         _max_atime:Optional[int] = None
         _inodes:Optional[int] = None
@@ -95,10 +95,18 @@ class GarbageGraph:
             @property
             def substitutable(_self):
                 if _self._substitutable is None:
-                    _self._substitutable = bool(self.store.query_substitutable_paths(
+                    _self._substitutable = _self.valid and bool(self.store.query_substitutable_paths(
                         {libstore.StorePath(_self.path)},
                     ))
                 return _self._substitutable
+
+            @property
+            def valid(_self):
+                return _self.nar_size is not None
+
+            @property
+            def size(_self):
+                return _self.nar_size if _self.nar_size is not None else _self.fs_size
 
             @property
             def score(_self):
@@ -110,24 +118,24 @@ class GarbageGraph:
                 if penalize_inodes is not None:
                     s -= penalize_inodes * _self.inodes_score
                 if penalize_size is not None:
-                    s -= penalize_size * _self.nar_size_score
+                    s -= penalize_size * _self.size_score
                 return s
 
             if limit_unit == QuantityUnit.BYTES:
                 @property
                 def limit_measurement(_self):
-                    return _self.nar_size
+                    return _self.size
 
                 @property
                 def inodes_score(_self):
-                    # divide by nar_size - we want to free many inodes before
-                    # we hit the limit, and the limit is based on nar_size.
-                    # add one to nar_size to avoid zero-division
-                    return _self.inodes / (_self.nar_size+1)
+                    # divide by size - we want to free many inodes before
+                    # we hit the limit, and the limit is based on size.
+                    # add one to size to avoid zero-division
+                    return _self.inodes / (_self.size+1)
 
                 @property
-                def nar_size_score(_self):
-                    return _self.nar_size
+                def size_score(_self):
+                    return _self.size
             elif limit_unit == QuantityUnit.INODES:
                 @property
                 def limit_measurement(_self):
@@ -138,11 +146,11 @@ class GarbageGraph:
                     return _self.inodes
 
                 @property
-                def nar_size_score(_self):
+                def size_score(_self):
                     # divide by inodes - we want to free many bytes before
                     # we hit the limit, and the limit is based on inodes.
                     # add one to inodes to avoid zero-division
-                    return _self.nar_size / (_self.inodes+1)
+                    return _self.size / (_self.inodes+1)
             else:
                 raise ValueError(f"Don't know how to measure {limit_unit!r}")
 
@@ -162,13 +170,13 @@ class GarbageGraph:
             action=libstore.GCAction.GCReturnDead,
         )
 
-        self.invalid_paths = set()
+        self.very_invalid_paths = set()
 
         def store_path_or_none(p):
             try:
                 return libstore.StorePath(path_split(str(p))[1])
             except RuntimeError:
-                self.invalid_paths.add(p)
+                self.very_invalid_paths.add(p)
                 return None
 
         garbage_store_path_set = {
@@ -189,25 +197,29 @@ class GarbageGraph:
         for store_path in reversed(garbage_store_paths_sorted):
             try:
                 path_info = self.store.query_path_info(store_path)
-            except RuntimeError:
-                self.invalid_paths.add(path_join(
-                    _nix_store_path,
-                    str(store_path),
-                ))
-            else:
+                str_path = str(path_info.path)
                 node_data = self.StorePathNode(
-                    str(path_info.path),
+                    str_path,
                     path_info.nar_size,
                 )
-                node_index = self.graph.add_node(node_data)
-                self.path_index_mapping[str(path_info.path)] = node_index
+            except RuntimeError:
+                path_info = None
+                str_path = str(store_path)
+                node_data = self.StorePathNode(
+                    str_path,
+                    None,
+                )
 
+            node_index = self.graph.add_node(node_data)
+            self.path_index_mapping[str_path] = node_index
+
+            if path_info is not None:
                 for ref_sp in path_info.references:
                     ref_node_index = self.path_index_mapping.get(str(ref_sp))
                     if ref_node_index == node_index:
                         logger.debug(
                             "omitting self-referencing edge from path %s",
-                            str(path_info.path),
+                            str_path,
                         )
                     elif ref_node_index is not None:
                         self.graph.add_edge(
@@ -222,13 +234,13 @@ class GarbageGraph:
         if _gc_keep_derivations or _gc_keep_outputs:
             logger.info("populating output-drv or drv-output edges")
             for path, idx in self.path_index_mapping.items():
-                if path.endswith(".drv"):
+                if self.graph[idx].valid and path.endswith(".drv"):
                     try:
                         for output in self.store.query_derivation_outputs(
                             libstore.StorePath(path),
                         ):
                             output_idx = self.path_index_mapping.get(str(output))
-                            if output_idx is not None:
+                            if output_idx is not None and self.graph[output_idx].valid:
                                 if _gc_keep_derivations:
                                     self.graph.add_edge(
                                         output_idx,
@@ -252,6 +264,7 @@ class GarbageGraph:
             logger.info("bulk querying path substitutability")
             substitutable_paths = self.store.query_substitutable_paths({
                 libstore.StorePath(self.graph[i].path) for i in pseudo_root_idxs
+                if self.graph[i].valid
             })
             for i in pseudo_root_idxs:
                 self.graph[i]._substitutable = (

--- a/nix_heuristic_gc/graph.py
+++ b/nix_heuristic_gc/graph.py
@@ -47,6 +47,7 @@ class GarbageGraph:
         store:libstore.Store,
         limit_unit:QuantityUnit,
         executor:Executor=NaiveExecutor(),
+        penalize_invalid:Optional[float]=None,
         penalize_substitutable:Optional[float]=None,
         penalize_drvs:Optional[float]=None,
         penalize_inodes:Optional[float]=None,
@@ -111,6 +112,8 @@ class GarbageGraph:
             @property
             def score(_self):
                 s = float(_self.max_atime)
+                if penalize_invalid is not None and not _self.valid:
+                    s -= penalize_invalid
                 if penalize_drvs is not None and _self.path.endswith(".drv"):
                     s -= penalize_drvs
                 if penalize_substitutable is not None and _self.substitutable:

--- a/tests/test_graph.py
+++ b/tests/test_graph.py
@@ -102,9 +102,9 @@ def test_basic_init(
             mock.call("/nix/store/cccccccccccccccccccccccccccccccc-ccc-3.3.3"),
             mock.call("/nix/store/dddddddddddddddddddddddddddddddd-ddd-4.4.4"),
             mock.call("/nix/store/55555555555555555555555555555555-555-5.5.5"),
+            mock.call("/nix/store/77777777777777777777777777777777-777-7.7.7"),
         ]
 
-        assert garbage_graph.invalid_paths == {
+        assert garbage_graph.very_invalid_paths == {
             "/nix/store/666-6.6.6",
-            "/nix/store/77777777777777777777777777777777-777-7.7.7",
         }

--- a/tests/test_nix_heuristic_gc.py
+++ b/tests/test_nix_heuristic_gc.py
@@ -19,6 +19,6 @@ def test_empty_local_store(caplog, dry_run, unit):
         threads=0,
     )
 
-    assert "requesting deletion of 0 store paths, total nar_size 0 bytes, 0 inodes" in caplog.text
+    assert "requesting deletion of 0 store paths, total size 0 bytes, 0 inodes" in caplog.text
     assert ("(not) requesting deletion" in caplog.text) is dry_run
     assert "ran out of zero-reference paths to remove" in caplog.text

